cpp/include/raft/neighbors/brute_force-inl.cuh

```diff
@@ -449,5 +450,31 @@ void search(raft::resources const& res,
   raft::neighbors::detail::brute_force_search<T, IdxT>(res, idx, queries, neighbors, distances);
 }
 
+/**
+ * @brief Brute Force search with filter using the constructed index.
+ *
+ * @tparam       T data element type
+ * @tparam       IdxT type of the indices
+ *
+ * @param[in] res raft resources
+ * @param[in] idx brute force index
+ * @param[in] queries a device matrix view to a row-major matrix [n_queries, index->dim()]
+ * @param[in] sample_filter a device filter function that green lights samples for a given query
+ * @param[out] neighbors a device matrix view to the indices of the neighbors in the source dataset
+ * [n_queries, k]
+ * @param[out] distances a device matrix view to the distances to the selected neighbors [n_queries,
+ * k]
+ */
+template <typename T, typename IdxT>
+void search_with_filtering(raft::resources const& res,
```

Comment:
Since we're adding this new and not updating an existing API, I think we should implement this the way we'd like to implement it going forward and not the way we've been doing in the past- let's just add an `std::optional` mask to the non-filtered search and then dispatch as appropriate.

Reply:
OK, do you wanna keep the `search_with_filtering` or merge to `search` entirely via `std::optional`?